File: C++/number_compliment.cpp
#include <iostream>
#include "./number_compliment.h"

using namespace std;

int Solution::findComplement(int num) {
    int temp = num;
    // if we use int here, the result will be different....
    unsigned int bitmask = 0xFFFFFFFF;
    while ( !(temp & (0x80000000)) ) {
        bitmask >>= 1;   // This operation will be affacted negative number: It will use "1" as the padding!
        temp <<= 1;  // This operation won't be affacted by negative number: It always use "0" as the padding.
    }
    return (~num) & (bitmask);
}

int main() {
    Solution* sol = new Solution();
    int output = sol->findComplement(1);
    delete sol;
    cout << output << "\n";
    return 0;
}

File: Python2/number_compliment.py
class Solution(object):
    def findComplement(self, num):
        bitmask = 0xFFFFFFFF  # all bits are "1"
        temp = num
        # find the number of leading zero bits we need to filter out
        while not temp & 0x80000000:
            # Because we will do AND operation. The zero bits will be used to filter the leading "1" bits after inverting the bits
            # On the contrary, the "1" bits won't affect the original value
            bitmask >>= 1
            # this is for knowing when we should stop: reaching the first non-zero bit
            temp <<= 1
        return ~(num) & bitmask


if __name__ == '__main__':
    sol = Solution()
    print sol.findComplement(1)
